File: app/api/src/main/kotlin/com/wallet/clover/api/service/AuthService.kt
package com.wallet.clover.api.service

import com.wallet.clover.api.dto.Auth
import com.wallet.clover.api.entity.user.UserEntity
import com.wallet.clover.api.entity.auth.RefreshTokenEntity
import com.wallet.clover.api.common.UserDefaults
import com.wallet.clover.api.repository.user.UserRepository
import com.wallet.clover.api.repository.auth.RefreshTokenRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDateTime

@Service
class AuthService(
    private val userRepository: UserRepository,
    private val jwtService: JwtService,
    private val refreshTokenRepository: RefreshTokenRepository,
    private val tokenBlacklistService: TokenBlacklistService
) {
    
    // TODO: 실제 JWT 라이브러리 구현 완료 후 테스트 필요
    // TODO: 프로덕션 배포 시 Secret Key를 환경변수로 변경
    
    /**
     * 로그인: Access Token + Refresh Token 발급
     */
    @Transactional
    suspend fun login(ssoQualifier: String): Auth.LoginResponse {
        val user = userRepository.findBySsoQualifier(ssoQualifier)
            ?: userRepository.save(
                UserEntity(
                    ssoQualifier = ssoQualifier,
                    age = UserDefaults.DEFAULT_AGE,
                    locale = UserDefaults.DEFAULT_LOCALE
                )
            )
        
        val accessToken = jwtService.generateAccessToken(user.id!!)
        val refreshToken = jwtService.generateRefreshToken(user.id)
        
        // Refresh Token DB 저장
        refreshTokenRepository.save(
            RefreshTokenEntity(
                userId = user.id,
                token = refreshToken,
                expiresAt = jwtService.getRefreshTokenExpiry()
            )
        )
        
        return Auth.LoginResponse(
            accessToken = accessToken,
            refreshToken = refreshToken,
            user = user
        )
    }
    
    /**
     * Refresh: Access Token 갱신
     */
    suspend fun refresh(refreshToken: String): Auth.RefreshResponse {
        // 1. Refresh Token 검증
        val userId = jwtService.validateRefreshToken(refreshToken)
        
        // 2. DB에서 Refresh Token 확인
        val storedToken = refreshTokenRepository.findByToken(refreshToken)
            ?: throw IllegalArgumentException("Invalid refresh token")
        
        // 3. 만료 확인
        if (storedToken.expiresAt.isBefore(LocalDateTime.now())) {
            refreshTokenRepository.deleteByToken(refreshToken)
            throw IllegalArgumentException("Refresh token expired")
        }
        
        // 4. 새 Access Token 발급
        val newAccessToken = jwtService.generateAccessToken(userId)
        
        return Auth.RefreshResponse(accessToken = newAccessToken)
    }
    
    /**
     * Logout: 토큰 무효화
     */
    suspend fun logout(accessToken: String, refreshToken: String) {
        // 1. Access Token을 Blacklist에 추가
        tokenBlacklistService.addToBlacklist(accessToken)
        
        // 2. Refresh Token 삭제
        refreshTokenRepository.deleteByToken(refreshToken)
    }
}

File: apps/backend/src/auth/auth.service.ts
import { Injectable, UnauthorizedException } from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import { UsersService } from '../users/users.service';
import { PrismaService } from '../prisma/prisma.service';

@Injectable()
export class AuthService {
  constructor(
    private readonly usersService: UsersService,
    private readonly jwtService: JwtService,
    private readonly prisma: PrismaService,
  ) {}

  async login(ssoQualifier: string, email?: string) {
    const user = await this.usersService.findOrCreateBySsoQualifier(
      ssoQualifier,
      email,
    );

    const accessToken = this.generateAccessToken(user.id);
    const refreshToken = this.generateRefreshToken(user.id);

    // Refresh Token DB 저장
    await this.prisma.refreshToken.create({
      data: {
        userId: user.id,
        token: refreshToken,
        expiresAt: new Date(Date.now() + 7 * 24 * 60 * 60 * 1000),
      },
    });

    return {
      accessToken,
      refreshToken,
      user,
    };
  }

  async refresh(refreshToken: string) {
    try {
      // 1. Refresh Token 검증
      const payload = this.jwtService.verify(refreshToken);
      const userId = BigInt(payload.sub);

      // 2. DB에서 유효한 토큰인지 확인
      const storedToken = await this.prisma.refreshToken.findUnique({
        where: { token: refreshToken },
      });

      if (!storedToken || storedToken.expiresAt < new Date()) {
        if (storedToken) {
          await this.prisma.refreshToken.delete({
            where: { id: storedToken.id },
          });
        }
        throw new UnauthorizedException(
          '유효하지 않거나 만료된 리프레시 토큰입니다',
        );
      }

      // 3. 새 Access Token 발급
      const accessToken = this.generateAccessToken(userId);

      return { accessToken };
    } catch (_e) {
      throw new UnauthorizedException('토큰 갱신에 실패했습니다');
    }
  }

  async logout(accessToken: string, refreshToken: string) {
    // 1. Access Token을 블랙리스트에 추가
    try {
      const payload = this.jwtService.decode(accessToken);
      if (payload && payload.exp) {
        await this.prisma.tokenBlacklist.create({
          data: {
            token: accessToken,
            expiresAt: new Date(payload.exp * 1000),
          },
        });
      }
    } catch (_e) {
      // Decode 실패 시 무시
    }

    // 2. Refresh Token 삭제
    await this.prisma.refreshToken.deleteMany({
      where: { token: refreshToken },
    });
  }

  private generateAccessToken(userId: bigint): string {
    return this.jwtService.sign(
      { sub: userId.toString() },
      { expiresIn: '1h' },
    );
  }

  private generateRefreshToken(userId: bigint): string {
    return this.jwtService.sign(
      { sub: userId.toString() },
      { expiresIn: '7d' },
    );
  }
}
